### PR TITLE
Disconnect reason

### PR DESCRIPTION
**Pull request by halyssonJr**
## Summary 
PR aims to add a disconnect reason, an explicit reason for your device disconnect, this way the user will collect the event's argument and check what happens. It was tested using the most common reason ( wrong password / SSID, manual disconnection, and lost AP connection). Please, check:
- https://github.com/WildernessLabs/Meadow.Contracts/pull/195

##  Result

```
Meadow StdOut: Run...
Meadow StdOut: wifi.IsConnected = False
Meadow StdOut: Connecting to WiFi Network
Meadow StdOut: Sending command to connect
Meadow StdOut: Sending ESP command. destination:1, function:2
Meadow StdOut: Wifi InvokeEvent NetworkDisconnectedEvent returned CompletedOk
Meadow StdOut: Disconnect Event, Reason 15
Meadow StdOut: Wifi_NetworkDisconnected event, Reason: Wrong password. Please check your WiFi credentials
Meadow StdOut: wifi.IsConnected = False
Meadow StdOut: ESP Ioctl returned success: CannotConnectToAccessPoint
Meadow StdOut: SendingCommand returned: CannotConnectToAccessPoint
Meadow StdOut: Error connecting to access point: Connection failed: CannotConnectToAccessPoint

Meadow StdOut: Run...
Meadow StdOut: wifi.IsConnected = False
Meadow StdOut: Connecting to WiFi Network
Meadow StdOut: Sending command to connect
Meadow StdOut: Sending ESP command. destination:1, function:2
Meadow StdOut: Wifi InvokeEvent NetworkDisconnectedEvent returned CompletedOk
Meadow StdOut: Disconnect Event, Reason 201
Meadow StdOut: Wifi_NetworkDisconnected event, Reason: Access Point not found. Please check your WiFi credentials or may be out of range of the access point
Meadow StdOut: wifi.IsConnected = False
Meadow StdOut: ESP Ioctl returned success: CannotConnectToAccessPoint
Meadow StdOut: SendingCommand returned: CannotConnectToAccessPoint
Meadow StdOut: Error connecting to access point: Connection failed: CannotConnectToAccessPoint

Meadow StdOut: Waiting for wifi to connect loop...
Meadow StdOut: wifi.IsConnected = True
Meadow StdOut: Waiting for wifi to connect loop...
Meadow StdOut: wifi.IsConnected = True
Meadow StdOut: Wifi InvokeEvent NetworkDisconnectedEvent returned CompletedOk
Meadow StdOut: Disconnect Event, Reason 200
Meadow StdOut: Wifi_NetworkDisconnected event, Reason: Close connection from Access Point
```

### CODE REVIEW

File: source/implementations/f7/Meadow.F7/Devices/F7EthernetNetworkAdapter.cs [hunk 1]
```diff
@@ -71,7 +71,7 @@ protected void InvokeEvent(EthernetFunction eventId, StatusCodes statusCode, byt
             case EthernetFunction.NetworkDisconnectedEvent:
                 Resolver.Log.Trace("Ethernet disconnected event triggered!");
 
-                RaiseNetworkDisconnected();
+                RaiseNetworkDisconnected(null);
```
Comment (adrianstevens) [inline source/implementations/f7/Meadow.F7/Devices/F7EthernetNetworkAdapter.cs:74]: And a NetworkDisconnectionEventArgs here?

File: source/implementations/f7/Meadow.F7/Devices/Esp32Coprocessor/Encoders.cs [hunk 2]
```diff
@@ -811,18 +811,20 @@ public static byte[] EncodeDisconnectEventData(MessagePayloads.DisconnectEventDa
...
             byte[] buffer = new byte[length];
             Array.Clear(buffer, 0, buffer.Length);
-            buffer[offset] = disconnectEventData.Retrying;
+            buffer[offset] = disconnectEventData.SsidLength;
             offset += 1;
-            EncodeInt32(disconnectEventData.RetriesRemaining, buffer, offset);
-            offset += 4;
-            EncodeUInt32(disconnectEventData.Reason, buffer, offset);
+            Array.Copy(buffer, offset, disconnectEventData.Bssid, 0, 6);
+            offset += (int) 6;
```
Comment (adrianstevens) [inline source/implementations/f7/Meadow.F7/Devices/Esp32Coprocessor/Encoders.cs:824]: isn't offset an int here? I don't think you need the cast

File: source/implementations/f7/Meadow.F7/Devices/Esp32Coprocessor/Encoders.cs [hunk 3]
```diff
@@ -836,11 +838,19 @@ public static MessagePayloads.DisconnectEventData ExtractDisconnectEventData(byt
         {
             DisconnectEventData disconnectEventData = new MessagePayloads.DisconnectEventData();
 
-            disconnectEventData.Retrying = buffer[offset];
+            for (int index = 0; (buffer[index + offset] != 0) && (index < (33 - 1)); index++)
+            {
+                disconnectEventData.Ssid += Convert.ToChar(buffer[index + offset]);
+            }
+            offset += (int) 33;
```
Comment (adrianstevens) [inline source/implementations/f7/Meadow.F7/Devices/Esp32Coprocessor/Encoders.cs:845]: and here

File: source/implementations/f7/Meadow.F7/Devices/Esp32Coprocessor/Encoders.cs [hunk 4]
```diff
@@ -836,11 +838,19 @@ public static MessagePayloads.DisconnectEventData ExtractDisconnectEventData(byt
...
+            disconnectEventData.SsidLength = buffer[offset];
             offset += 1;
-            disconnectEventData.RetriesRemaining = ExtractInt32(buffer, offset);
-            offset += 4;
-            disconnectEventData.Reason = ExtractUInt32(buffer, offset);
+            disconnectEventData.Bssid = new byte[6];
+            Array.Copy(buffer, offset, disconnectEventData.Bssid, 0, 6);
+            offset += (int) 6;
```
Comment (adrianstevens) [inline source/implementations/f7/Meadow.F7/Devices/Esp32Coprocessor/Encoders.cs:850]: and here

File: source/implementations/f7/Meadow.F7/Devices/F7CellNetworkAdapter.cs [hunk 5]
```diff
@@ -121,7 +121,7 @@ protected void InvokeEvent(CellFunction eventId, StatusCodes statusCode, byte[]
 
                 ResetCellTempData();
 
-                RaiseNetworkDisconnected();
+                RaiseNetworkDisconnected(null);
```
[the remark this answers is not in the document]
Comment (halyssonJr) [reply source/implementations/f7/Meadow.F7/Devices/F7CellNetworkAdapter.cs:124]: Considering this argument precedes a disconnection reason or value that can help the user know what exactly occurred, I think we should keep NULL until we reunite the ETHERN and Cell reasons to disconnect.

**Review by adrianstevens — APPROVED** [review]

**Pull request merged**